### PR TITLE
Add sparse_attention mask ,test=develop

### PR DESCRIPTION
**Pull request by Liu-xiandong**
### PR types

Function optimization
### PR changes

APIs
### Describe

Add key_padding_mask and attn_mask in sparse_attention Api

1.Key padding mask is a tensor with dimensions [batch_size, seq_len], and attention mask is a tensor with dimensions [seq_len, seq_len]. The data types of the two masks are consistent with Q, K, and V, which are float32 or float64. If the value in Mask is 0, it means that the position needs to be masked.

2.The changed files are mainly paddle/fluid/operators/sparse_attention_op.cu and python/paddle/fluid/tests/unittests/test_sparse_attention_op.py.  sparse_attention has three parts: **sddmm, softmax, and dsd**. Adding the mask operation only needs to modify the  **softmax**. It has no effect on the other two parts. In addition, in order to test the mask function, related tests has been added.

3.Since the CI platform does not have a  machine with CUDA11.3, the local results are shown as follows:
02dcb9f8ad7f36a35e2c06aa82fd607e
image

### CODE REVIEW

File: paddle/fluid/operators/sparse_attention_op.cu [hunk 1]
```diff
@@ -72,24 +72,32 @@ __global__ void BlockSparseSoftmaxForward(T* softmax, const T* src, T scale,
     const int cur_block_nnz =
         layout_rowptr[cur_block_row + 1] - layout_rowptr[cur_block_row];
 
-    T srcdata[(BlockSize * BlockNnzMax + WarpSize - 1) / WarpSize];
-    T attndata[(BlockSize * BlockNnzMax + WarpSize - 1) / WarpSize];
-
-    // read kp mask
-    T cur_kp_mask = (kp_mask == nullptr) ? 0 : kp_mask[cur_row];
+    T srcdata[(BlockSize * BlockNnzMax + WarpSize - 1) / WarpSize] = {0};
```
Comment (xingfeng01) [inline paddle/fluid/operators/sparse_attention_op.cu:75]: 不是很确定会把向量中的所有变量都初始化为0吗

File: python/paddle/fluid/tests/unittests/test_sparse_attention_op.py [hunk 2]
```diff
@@ -280,6 +459,7 @@ def setUp(self):
         self.shape = (2, 2, 8, 4)
         self.blocksize = 2
         self.dtype = 'float32'
+        self.use_mask = False
```
Comment (xingfeng01) [inline python/paddle/fluid/tests/unittests/test_sparse_attention_op.py:460]: 有加入 mask=true 的测试吗

File: python/paddle/fluid/tests/unittests/test_sparse_attention_op.py [hunk 3]
```diff
@@ -254,20 +400,53 @@ def test_dygraph(self):
         query = np.random.random(self.shape).astype(self.dtype)
         key = np.random.random(self.shape).astype(self.dtype)
         value = np.random.random(self.shape).astype(self.dtype)
+        # init mask tensor
+        key_padding_mask_shape = (self.shape[0], self.shape[2])
+        attn_mask_shape = (self.shape[2], self.shape[2])
+        key_padding_mask = np.random.randint(0, 2, size=key_padding_mask_shape)
+        #key_padding_mask = np.array([[1,1,1,1,1,1,1,0]])
```
Comment (xingfeng01) [inline python/paddle/fluid/tests/unittests/test_sparse_attention_op.py]: 删除掉临时的注释吧

Comment (Liu-xiandong) [reply python/paddle/fluid/tests/unittests/test_sparse_attention_op.py]: OK

File: python/paddle/fluid/tests/unittests/test_sparse_attention_op.py (hunk 2 above)
Comment (Liu-xiandong) [reply python/paddle/fluid/tests/unittests/test_sparse_attention_op.py:460]: 203行和300行，有mask=true的测试

File: paddle/fluid/operators/sparse_attention_op.cu (hunk 1 above)
Comment (Liu-xiandong) [reply paddle/fluid/operators/sparse_attention_op.cu:75]: 做了一些试验，可以初始化为0。然后后面有一个读数的过程，将srcdata中的数据初始化为负无穷。所以这个地方的正确性是可以保证的。

**Review by xingfeng01 — APPROVED** [review]

**Review by zkh2016 — APPROVED** [review]

**Review by lanxianghit — APPROVED** [review]

**Review by dingjiaweiww — APPROVED** [review]
LGTM
